### PR TITLE
Developer setup instructions

### PR DESCRIPTION
**Pull request by tiagojsag**
These changes should cover the necessary steps to run the microservices under linux.

As someone who can probably do these steps in their sleep, my opinion is the least valuable when it comes to "is this detailed enough for a new user?". In that sense, I welcome everyone's feedback on ways we can improve this for a newly arrived user.

Missing:
- Whenever we replace Control Tower with something else, we need to see how the dev setup will look. This will probably be done after this PR is merged.

### CODE REVIEW

**Review by ethanroday — CHANGES_REQUESTED** [review]
This is a great start. Few things I see missing:
- Mac instructions (as you've noted already)
- More info about common errors and how to overcome them (e.g. you installed MongoDB but forgot to start the service)
- Lots more info about the deployment process and getting to production. Specifically:
    - Environments - what environments exist and what's the process for going from one to the other?
    - Automatic checks - what gets run and where do I see the results?
    - Jenkins - where do I go to see status? what should I do if a build fails?
    - Testing in staging - what's best practice for hitting these endpoints in staging?
    - Smoke tests - instructions to add smoke tests for new endpoints

**Comment by tiagojsag** [comment]
@ethanroday thanks for the feedback.

Mac instructions are already included - my last commit made some small improvements for Mac users, but overall I was able to run a MS in both modes on a Mac using the steps already covered. Let me know if there's anything I missed

On errors: I added a few that came to mind, but let me know if you have any more suggestions we should add.

I'll work on the remaining stuff.

**Comment by tiagojsag** [comment]
The latest commits should cover the remaining topics. let me know your thoughts.

File: source/includes/dev/_microservice_dev_guide.md
Comment (ikas) [inline source/includes/dev/_microservice_dev_guide.md]: Please add a reference here to the section in the developer docs about smoke tests: https://resource-watch.github.io/doc-api/developer.html#api-smoke-tests

File: source/includes/dev/_microservice_dev_guide.md
Comment (ikas) [inline source/includes/dev/_microservice_dev_guide.md:294]: Another "recurring" issue that might cause builds in Jenkins to fail is related to Docker networks - the error message says something like "Cannot obtain IPv4 address to create Docker network" and the resolution is pruning the networks.

**Review by ikas — APPROVED** [review]
LGTM overall, small comments added :+1:

File: source/includes/dev/_microservice_dev_guide.md
Comment (tiagojsag) [reply source/includes/dev/_microservice_dev_guide.md:294]: that's mentioned already

Comment (ikas) [reply source/includes/dev/_microservice_dev_guide.md:294]: Yeah it is, sorry I missed it! Resolving!

File: source/includes/dev/_microservice_dev_guide.md
Comment (ethanroday) [inline source/includes/dev/_microservice_dev_guide.md:164]: Love this section!

File: source/includes/dev/_microservice_dev_guide.md
Comment (ethanroday) [inline source/includes/dev/_microservice_dev_guide.md]:
```suggestion
- **Your microservice has user-related issues, even though you are providing a `Bearer` token**: Bearer tokens are processed by Control Tower, and transformed into a `loggedUser` JSON object that microservices expect. Either provide said object directly to your microservice, or route your request with the token through Control Tower.
```

File: source/includes/dev/_microservice_dev_guide.md
Comment (ethanroday) [inline source/includes/dev/_microservice_dev_guide.md:224]: This section is duplicated, no?

File: source/includes/dev/_microservice_dev_guide.md
Comment (ethanroday) [inline source/includes/dev/_microservice_dev_guide.md]: Would consider adding a little info here - we are in the process of renaming branches, etc.

File: source/includes/dev/_microservice_dev_guide.md
Comment (ethanroday) [inline source/includes/dev/_microservice_dev_guide.md]:
```suggestion
- `dev` for internal testing and development of new features. There are no guarantees of stability or data persistence. While it's not barred from public access, it's meant to be used only by developers working on the RW API code, for testing, debugging and experimentation.
```

File: source/includes/dev/_microservice_dev_guide.md
Comment (ethanroday) [inline source/includes/dev/_microservice_dev_guide.md]: Would take a firmer stance here and say that the process is to announce _at least_ 14 days in advance of the intended production deployment in the `general` channel in the WRI API Slack (and to contact Ethan Roday if you're not in that Slack workspace). In rare cases if a hotfix is needed to fix a breaking change in production, the 14-day lead time can be circumvented, but an announcement still must be made.

File: source/includes/dev/_microservice_dev_guide.md
Comment (ethanroday) [inline source/includes/dev/_microservice_dev_guide.md]: Not clear whether I actually have to trigger a build here or if that happens automatically once the PR is merged.

**Comment by ethanroday** [comment]
Love the deployment instructions; very helpful. Couple of suggestions in addition to the small edits I just made:
1. Would be great to have a step-by-step overview of the deployment process before getting into the details. Something like this (though not sure where Jenkins fits in):
  - Make a feature branch
  - Write your code
  - Test your feature branch locally
  - Make a PR from feature to dev
  - Wait for Travis and approval, then merge and test in dev
  - Make a PR from dev to staging
  - Wait for Travis and approval, then merge and test in staging
  - Make an announcement about an upcoming deployment
  - Make a PR from staging to prod
  - Wait for Travis and approval, then merge and test in staging
2. Can we actually call out what the public URLs are for dev and staging? (Or did you do that and I missed it?)

File: source/includes/dev/_microservice_dev_guide.md
Comment (tiagojsag) [reply source/includes/dev/_microservice_dev_guide.md:224]: no, there are different ways to handle things for docker users vs native users.

File: source/includes/dev/_microservice_dev_guide.md
Comment (tiagojsag) [reply source/includes/dev/_microservice_dev_guide.md:164]: well show it some love by trying to remember if you ran into something that's not covered here :P Kidding aside, I struggled a bit to write it, as I don't run into these pitfalls anymore - I already have the whole API configured on my computer. Since you went through this process recently, if you have any suggestions of problems you faced, let me know.

**Comment by tiagojsag** [comment]
@ethanroday think I tackled everything, but let me know if I missed anything

File: source/includes/dev/_microservice_dev_guide.md
Comment (thomas-maschler) [inline source/includes/dev/_microservice_dev_guide.md]: True that there is no technical reason to use a different language, but it is highly discouraged. We want to keep the technical stack as tight as possible and not introduce technologies or programming languages which are unfamiliar to the core development team. Any introduction of a new technology must be well justified in an ADR. New services should always default to well known technology/ coding language already in use.

File: source/includes/dev/_microservice_dev_guide.md
Comment (thomas-maschler) [inline source/includes/dev/_microservice_dev_guide.md:49]: The final service will run inside a docker container. When using docker containers already during development you reduce the risk of having to justify yourself with _"But it worked on my machine"_ too often, because you forgot to add the required dependencies to the docker image before you deployed your code.

File: source/includes/dev/_microservice_dev_guide.md
Comment (thomas-maschler) [reply source/includes/dev/_microservice_dev_guide.md:164]: I had some issues at one point that a dependent service on my local machine was out of date.
Always remember to develop against the current/stable branch of CT and other services

File: source/includes/dev/_microservice_dev_guide.md
Comment (thomas-maschler) [reply source/includes/dev/_microservice_dev_guide.md:224]: What about the `mymachine` entry in /etc/hosts ?

File: source/includes/dev/_microservice_dev_guide.md
Comment (thomas-maschler) [inline source/includes/dev/_microservice_dev_guide.md]: I understand that there is some variation here due to legacy and projected specific decisions. But I think it is important to highlight again the best practice approach so that we can streamline this more in the future.

File: source/includes/dev/_microservice_dev_guide.md
Comment (thomas-maschler) [inline source/includes/dev/_microservice_dev_guide.md]: TYPO:  more often _than_ not

File: source/includes/dev/_microservice_dev_guide.md
Comment (thomas-maschler) [reply source/includes/dev/_microservice_dev_guide.md:224]: CT doesn't like endpoint changes. If you change routes in your microservice you might have to delete control tower and the associated docker data at ~/docker/control-tower to get things to work again

Comment (thomas-maschler) [reply source/includes/dev/_microservice_dev_guide.md:224]: You might run out of disc space due to too many docker images and data stored on you machine.
`docker system prune --all` removes all images, volumes and networks not currently in use and usually frees a lot of disk space this way. 
After running docker system prune, spinning up a docker container might take a little longer at first, b/c docker needs to download the images again.

**Review by thomas-maschler — CHANGES_REQUESTED** [review]

**Comment by thomas-maschler** [comment]
Thanks for putting this together in such detail.

Sections that are still missing are 
- k8s secrets
- What to do when you need access to a database on AWS

Which leads to the next big topic: Infrastructure provisioning. @ethanroday do we plan to write a section on this as well?

File: source/includes/dev/_microservice_dev_guide.md
Comment (tiagojsag) [reply source/includes/dev/_microservice_dev_guide.md:224]: - On CT not liking endpoint changes: that should be handled by the cron I recently refactored. It is mentioned that, if you use CT, you should also run the cron.
- Ouf of disc space: IMO that's out of scope. As a rule of thumb, the docs should focus on what's specific to the API. In these sections, I've ignored that rule somewhat, for the sake of practicality, but the behavior you're describing is purely docker related, and can happen in any other docker project.

File: source/includes/dev/_microservice_dev_guide.md
Comment (tiagojsag) [reply source/includes/dev/_microservice_dev_guide.md]: IMO that's a temporary process we're carrying out that, by the time anyone reads these docs, will be mostly done.

**Comment by tiagojsag** [comment]
IMO we are going beyond what I would call "developer documentation" and going into "sysadmin/devops documentation". While we never really formalized the distinction between the two roles, we just need to compare the list of Jenkins users with the list of AWS users to see that a bunch of people develop for the api, but only a subset of those have access to AWS, k8s config and so on - as it should be.

For clarity for both groups, I think we should perhaps try to crystallize what are the responsibilities/permissions each role has and doesn't have, so that the documentation can reflect those. So, picking up on what @thomas-maschler requested on secrets, these docs probably should cover how a dev can use a k8s secret in their microservice, but not necessarily how to add said secret to the cluster - that would be a responsibility of a sysadmin. Likewise, database access is a gray area between the two (some applications allow devs to access live production databases, others allow access to snapshots, others don't allow it at all).

**Comment by ethanroday** [comment]
Sorry for the delay in responding here. I would say that [GTC-742](https://gfw.atlassian.net/browse/GTC-742) encapsulates both how to use the deployment pipeline (as a developer) and how to manage the development pipeline (as an administrator).
@tiagojsag as you have mentioned in the past, the whole idea of these docs is that if you were hit by a bus we would still be able to move forward 😄

**Comment by ethanroday** [comment]
Couple of small things to add, but I will make them a separate issue so we can merge this PR:
- Instructions on how to set up SSH for Bastion host (I already have these written down somewhere)
- Instructions on how to apply changes manually for the TF k8s config

**Review by ethanroday — APPROVED** [review]

**Comment by tiagojsag** [comment]
I focused the docs on what's specific to the RW API, and left out instruction on how to use 3rd party tools like ssh, terraform, aws or kubectl CLIs. I feel like not only are we adding more work for ourselves now, but also more work for us in the future - say terraform decides to change their CLI syntax, we now have outdated docs that need updating. 

Also, not only won't our docs be as complete and detailed as the "real" ones, but we can't really ensure things like cross-OS compatibility. So my focus was to cover what's exclusive to the RW API, with some pointers to 3rd party resources where the reader can learn how to create an SSH tunnel or apply terraform changes. 

I understand that it's convenient to "bend" this rule once in a while, but I wanted to clarify the motivation for the current approach, and also point it out as a potential pitfall we may want to keep in mind when extending these docs in the future.

**Pull request merged**
